### PR TITLE
Added support for configuring route options with LinkGenerator

### LINKED ISSUE #547: Link generation: Expose options to configure link generation at the call site and globally
- fully qualified URLs
- Casing
- trailing slashes

### PR DESCRIPTION
**Pull request by kichalla**
[Fixes #547] Link generation: Expose options to configure link generation at the call site and globally

### CODE REVIEW

File: src/Microsoft.AspNetCore.Routing.Abstractions/LinkGenerator.cs [hunk 1]
```diff
@@ -1,24 +1,13 @@
...
 {
     public abstract class LinkGenerator
     {
-        public abstract bool TryGetLink(
-            HttpContext httpContext,
-            IEnumerable<Endpoint> endpoints,
-            RouteValueDictionary explicitValues,
-            RouteValueDictionary ambientValues,
-            out string link);
+        public abstract bool TryGetLink(LinkGeneratorContext context, out string link);
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing.Abstractions/LinkGenerator.cs]: 👍 I'm pro a context object. More extensible, and there were getting too many parameters

File: src/Microsoft.AspNetCore.Routing/DefaultLinkGenerator.cs [hunk 2]
```diff
@@ -11,54 +11,60 @@
...
+        private readonly static char[] UrlQueryDelimiters = new char[] { '?', '#' };
         private readonly MatchProcessorFactory _matchProcessorFactory;
         private readonly ObjectPool<UriBuildingContext> _uriBuildingContextPool;
+        private readonly RouteOptions _options;
         private readonly ILogger<DefaultLinkGenerator> _logger;
 
         public DefaultLinkGenerator(
             MatchProcessorFactory matchProcessorFactory,
             ObjectPool<UriBuildingContext> uriBuildingContextPool,
+            IOptions<RouteOptions> options,
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Routing/DefaultLinkGenerator.cs]: options -> routeOptions, just to be explicit

File: test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs [hunk 3]
```diff
@@ -239,15 +259,259 @@ public void GetLink_Success_AmbientValues()
...
+            var link = linkGenerator.GetLink(
+                new LinkGeneratorContext
+                {
+                    Endpoints = new[] { endpoint },
+                    ExplicitValues = context.ExplicitValues,
+                    AmbientValues = context.AmbientValues
+                });
+
+            // Assert
+            Assert.Equal("/Home/Index?ShowStatus=True&INFO=DETAILED", link);
```
Comment (JamesNK) [inline test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs]: query string is not lowercase?

Comment (kichalla) [reply test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs]: It is lower cased only when LowercaseUrls is also true

Comment (JamesNK) [reply test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs]: Is that existing behavior?

Need to make sure that is documented on property

Comment (kichalla) [reply test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs]: Yeah, its an existing behavior (i know it feels weird) https://github.com/aspnet/Routing/blob/release/2.2/src/Microsoft.AspNetCore.Routing/RouteOptions.cs#L20

Comment (JamesNK) [reply test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs]: Ok! Copy over those properties xml docs

File: src/Microsoft.AspNetCore.Routing.Abstractions/LinkGenerator.cs (hunk 1 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing.Abstractions/LinkGenerator.cs]: I'm ok with making some tenative decisions here if we think it's an improvement. We still need to do design work on this area in preview1 - so I would suggest not making any expensive changes until we've had a proper design review.

**Comment by kichalla** [comment]
https://github.com/aspnet/Routing/commit/5f1631ab46e415cadf3a12ceef6c6164a06ac275

**Pull request closed without merging**